live-build/misc/live-build-hooks/81-upgrade-repository.binary

```diff
@@ -42,41 +58,50 @@ chroot binary bash -c "cd packages && xargs -a all-packages.list apt download"
 rename 's/\%3a/:/g' binary/packages/*.deb
 
 #
-# convert all-packages.list to a format appropriate for the 'Depends'
-# section of a package definition.
-#
-# Example input:
-#  bash=4.4.18-2ubuntu1
-#  bc=1.07.1-2
-# output:
-#  bash (= 4.4.18-2ubuntu1), bc (= 1.07.1-2)
+# Now we need to create the "delphix-entire" metapackage that will
+# depend on all of the packages that have been installed in the binary
+# chroot environment.
 #
-DEPENDS="$(tr '\n' ' ' <binary/packages/all-packages.list |
-	sed 's/ *$/)/' | sed 's/ /), /g' | sed 's/=/ (= /g')"
+cp config/hooks/template.ctl delphix-entire.ctl
+
+# shellcheck disable=SC2016
+DEPENDS=$(chroot binary \
+	dpkg-query -Wf '${Package} (=${Version}), ' | sed 's/, $//')
+
+sed -i "s/@@DEPENDS@@/$DEPENDS/" delphix-entire.ctl
+sed -i "s/@@VARIANT@@/$APPLIANCE_VARIANT/" delphix-entire.ctl
+sed -i "s/@@DATE@@/$(date '+%Y%m%dT%H%M%S')/" delphix-entire.ctl
+
+equivs-build delphix-entire.ctl
 
 #
-# build the delphix-entire meta-pacakge
+# Once we have the "delphix-entire" metapackage built, we need to
+# install it into the variant's binary chroot environment, so that the
+# appliance has this package installed. Otherwise, this package would
+# wind up in the upgrade repository, but not already installed in a
+# running appliance that hasn't yet be upgraded.
 #
-cp config/hooks/delphix-entire-template delphix-entire-control
-sed -i "s/@@DEPENDS@@/${DEPENDS}/" delphix-entire-control
-sed -i "s/@@DELPHIX_VERSION@@/1.0.0-$(date '+%Y%m%dT%H%M%S')/" \
-	delphix-entire-control
-equivs-build delphix-entire-control
-rm delphix-entire-control
+DEBS_DIRECTORY=$(mktemp -d -p /tmp tmp.debs.XXXXXXXXXX)
+
+cp ./*.deb "$DEBS_DIRECTORY"
+mv "$DEBS_DIRECTORY" "binary/$DEBS_DIRECTORY"
+chroot binary bash -c "dpkg -i $DEBS_DIRECTORY/*.deb"
+
+rm -rf "binary/$DEBS_DIRECTORY"
 
 #
-# Create the upgrade repository, put all the debs into it, and finally publish
-# it (into ~/.aptly/public by default).
-# We first clear any remnants from the previous run
+# Lastly, we generate the Aptly/APT repository using the directory full
+# of packages previously downloaded, and the delphix-entire package that
+# was just generated.
 #
-rm -rf "$HOME/.aptly"
-aptly repo create -distribution=bionic -component=dlpx upgrade-repo
-aptly repo add upgrade-repo binary/packages
-aptly repo add upgrade-repo delphix-entire*.deb
-aptly publish repo -skip-signing upgrade-repo
-rm delphix-entire*.deb
```

Comment:
perhaps remove the debs since we don't need them anymore?

Reply:
we can, but we already leave around a lot of other stuff, so I didn't think it was necessary to single out this specific build artifact.. e.g. we have the "binary" directory, "chroot" directory, etc.. but if you want, I can remove them.